src/drivers/kinetis/i2c.c

```diff
@@ -0,0 +1,310 @@
+                bestmult = mult;
+                besticr = icr;
+                besterr = abserr;
+            }
+        }
+        if (0 == besterr) {
+            break;
+        }
+    }
+    base->F = I2C_F_MULT(bestmult) | I2C_F_ICR(besticr);
+
+    /* configure stop-hold enable */
+    if(true == config.stophold) {
+        base->FLT |= I2C_FLT_SHEN_MASK;
+    }
+    else {
+        base->FLT &= I2C_FLT_SHEN_MASK;
+    }
+
+    /* configure glitch filter */
+    base->FLT |= I2C_FLT_FLT(config.filter_clocks);
+
+    /* enable i2c */
+    base->C1 |= I2C_C1_IICEN_MASK;
+
+    /* initialize the channels to be available */
+    for(int i = 0; i < I2C_NUM_DEVICES; i++) {
+        i2c_chs[i].status = I2C_AVAILABLE;
+    }
+}
+
+uint32_t i2c_send_seq(uint32_t ch_num, i2c_seq_t *seq, uint32_t seq_len, uint8_t *received_data, void (*callback)(void *), void *args) {
+    /* get the channel and the base from the globals */
+    volatile I2C_Channel *ch = &i2c_chs[ch_num];
+    I2C_Type *base = (I2C_Type *)i2c_bases[ch_num];
+
+    /* initialize the errno to zero */
+    uint8_t errno = 0;
+
+    /* error out if the peripheral is busy */
+    if(I2C_BUSY == ch->status) {
+        errno = -1;
+        goto exit;
+    }
+
+    /* initializing transfer */
+    ch->seq = seq;
+    ch->seq_end = seq + seq_len;
+    ch->received_data = received_data;
+    ch->status = I2C_BUSY;
+    ch->txrx = I2C_WRITING;
+    ch->callback = callback;
+    ch->args = args;
+
+    /* clear IICIF by writing a 1 to it */
+    base->S |= I2C_S_IICIF_MASK;
+
+    /* enable interrupts */
+    base->C1 = (I2C_C1_IICEN_MASK | I2C_C1_IICIE_MASK);
+
+    /* generate a start condition and set to transmit */
+    base->C1 |= (I2C_C1_MST_MASK | I2C_C1_TX_MASK);
+
+    if(base->S & I2C_S_ARBL_MASK) {
+        errno = -1;
+        goto i2c_send_seq_cleanup;
+    }
+
+    /* write the address */
+    base->D = (*ch->seq).addr;
+    ch->seq++;
+    goto exit;
+
+i2c_send_seq_cleanup:
+    /* generate a STOP, disable interrupts, change to RX */
+    base->C1 &= ~(I2C_C1_MST_MASK | I2C_C1_IICIE_MASK | I2C_C1_TX_MASK);
+    ch->status = I2C_ERROR;
+
+exit:
+    return errno;
+}
+
+uint32_t i2c_irq_handler(uint8_t ch_num) {
+#ifdef __KINETIS_I2C_BLOCKING_ISR
+    asm("cpsid i");
+#endif
+    /* get the channel and the base from the globals */
+    volatile I2C_Channel *ch = &i2c_chs[ch_num];
+    I2C_Type *base = (I2C_Type *)i2c_bases[ch_num];
+
+    /* initialize errno to zero */
+    int8_t errno = 0;
+
+    /* ensure that irq was from the current I2C module */
+    if(!(base->S & I2C_S_IICIF_MASK)) {
+        errno = -2;
+        goto i2c_isr_exit;
+    }
+
+    /* clear IICIF by writing a 1 to it */
+    I2C0->S |= I2C_S_IICIF_MASK;
+    
+    /* check if arbitration lost */
+    if(base->S & I2C_S_ARBL_MASK) {
+        base->S |= I2C_S_ARBL_MASK;
+        goto i2c_isr_error;
+    }
+
+    /* handle reading */
+    if(ch->txrx == I2C_READING) {
+        switch(ch->reads_ahead) {
```

Comment:
I assume this is how many reads are left in a sequence?  The channel struct is quite complex, you might want to document what all the variables do in the various ISR states.